source-mongodb/pull.go

```diff
@@ -140,30 +179,32 @@ type changeEvent struct {
 const resumePointGoneErrorMessage = "the resume point may no longer be in the oplog"
 const resumeTokenNotFoundErrorMessage = "the resume token was not found"
 
-func (c *capture) ChangeStream(ctx context.Context, client *mongo.Client, binding int, res resource, backfillFinishedAt *primitive.Timestamp, resumeToken bson.Raw) error {
+func (c *capture) ChangeStream(ctx context.Context, client *mongo.Client, binding int, res resource, state *resourceState) error {
 	var db = client.Database(res.Database)
 
 	var collection = db.Collection(res.Collection)
 
 	log.WithFields(log.Fields{
 		"database":           res.Database,
 		"collection":         res.Collection,
-		"resumeToken":        resumeToken,
-		"backfillFinishedAt": backfillFinishedAt,
+		"state":              state,
 	}).Info("listening on changes on collection")
 	var eventFilter = bson.D{{"$match", bson.D{{"$or",
-		bson.A{
-			bson.D{{"operationType", "delete"}},
-			bson.D{{"operationType", "insert"}},
-			bson.D{{"operationType", "update"}},
-			bson.D{{"operationType", "replace"}},
-		}}}}}
+	bson.A{
+		bson.D{{"operationType", "delete"}},
+		bson.D{{"operationType", "insert"}},
+		bson.D{{"operationType", "update"}},
+		bson.D{{"operationType", "replace"}},
+	}}}}}
+
 	var opts = options.ChangeStream().SetFullDocument(options.UpdateLookup)
-	if resumeToken != nil {
-		opts = opts.SetResumeAfter(resumeToken)
-	} else if backfillFinishedAt != nil {
-		opts = opts.SetStartAtOperationTime(backfillFinishedAt)
+	if state.StreamResumeToken != nil {
+		opts = opts.SetResumeAfter(state.StreamResumeToken)
+	} else if !state.BackfillStartedAt.IsZero() {
+		var startAt = &primitive.Timestamp{T: uint32(state.BackfillStartedAt.Unix()) - 1, I: 0}
```

Comment:
Do we need to go back in time any more than 1 second? I'm not sure how MongoDB streams work but am wondering if there is some "approximation" to when the oplog events are recorded that could be off by more than 1 second from our recorded wall clock time of when the backfill began, or even just drift between clocks of the server vs. connector.

Reply:
The 1 second here is only because `primitive.Timestamp` here only supports seconds precision, and I wanted to have the safety of one second as we have loss of precision from the original timestamp (I think lost precisions are rounded, so if we have >50 milliseconds, we end up rounding up a full second).

Reply:
I see - still wondering if we should have a longer safety factor though: For a case where MongoDB's notion of what time it is differs from what time our connector thinks it is a significant amount, to make sure that we don't miss any changes that happen right around the exact time that the backfill begins. It would be pretty unlikely, and would basically have to be a change in the first key in sorted order happening at almost the exact same time as the backfill starts, but maybe a ~5 minute buffer would be worth it to make that basically impossible.